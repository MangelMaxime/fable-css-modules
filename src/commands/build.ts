import type { Arguments, CommandBuilder } from "yargs";
import fg from "fast-glob";
import postcss from "postcss";
import { ClassCollector } from "../classCollector";
import { PostcssCollectorPlugin } from "../postcssCollectorPlugin";
import { getCssContent } from "../css";
import path from "path";

type Options = {
    source: string;
    outFile: string;
    internal : boolean
};

export const command: string = "$0 [source]";
export const desc: string = 'Generate bindings for all the CSS modules in <source> folder to <outFile> file';

export const builder: CommandBuilder<Options, Options> = (yargs) =>
    yargs
        .options({
            outFile: {
                describe: 'Output file where the binding will be written',
                type: 'string',
                default: 'CssModules.fs'
            },
            internal: {
                describe: 'Mark the generated module as internal',
                type: 'boolean',
                default: true
            }
        })
        .positional("source", {
            describe: 'Folder where the CSS modules are located',
            type: 'string',
            default: '.'
        });

export const handler = async (argv: Arguments<Options>) => {
    const sourceFolder = path.normalize(argv.source).replace(/\\/g, '/');
    const destinationFile = path.normalize(argv.outFile).replace(/\\/g, '/');

    const entries = await fg(`${sourceFolder}/**/*.module.scss`);

    const classCollector =
        new ClassCollector(
            process.cwd(),
            sourceFolder,
            destinationFile,
            argv.internal
        );

    for (const entry of entries) {
        // Transform from SASS to CSS
        const cssContent = await getCssContent(entry);

        // Register the postcss plugins
        const plugins = [
            PostcssCollectorPlugin({ collector: classCollector })
        ];

        // Run postcss to collect the classes
        await postcss(plugins).process(cssContent, { from: entry })
    }

    // Write the content to the destination file
    await classCollector.writeToFile();

    // Notify the user
    console.log("Generation completed");
};
