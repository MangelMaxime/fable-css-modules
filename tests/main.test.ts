import { expect, test } from 'vitest'
import { handler } from '../src/commands/build'
import { readFileSync } from "node:fs"
import pino from "pino"

test('adds 1 + 2 to equal 3', async () => {
    const logs = new Array<string>()
    const dest = {
        write: (data: string) => logs.push(data),
    }
    const logger = pino({}, dest)

    await handler({
        source: "fixtures/noCssModules",
        outFile: 'CssModules.fs',
        internal: true,
        _: [],
        $0: ''
    }, logger)

    expect(logs).length(2)
    expect(logs[0]).toContain("No classes found")
    expect(logs[1]).toContain("Generation completed")
})

test('contains 1 class if there is one CSS Module', async () => {
    await handler({
        source: "fixtures/oneCssModule",
        outFile: 'fixtures/oneCssModule/CssModules.fs',
        internal: true,
        _: [],
        $0: ''
    })

    const content = readFileSync("fixtures/oneCssModule/CssModules.fs").toString()
    expect(content).toMatchSnapshot()
})

test('contains 1 class per CSS Module', async () => {
    await handler({
        source: "fixtures/multipleCssModules",
        outFile: 'fixtures/multipleCssModules/CssModules.fs',
        internal: true,
        _: [],
        $0: ''
    })

    const content = readFileSync("fixtures/multipleCssModules/CssModules.fs").toString()
    expect(content).toMatchSnapshot()
})

test('support CSS/SASS/SCSS modules', async () => {
    await handler({
        source: "fixtures/checkSupportedFormat",
        outFile: 'fixtures/checkSupportedFormat/CssModules.fs',
        internal: true,
        _: [],
        $0: ''
    })

    const content = readFileSync("fixtures/checkSupportedFormat/CssModules.fs").toString()
    expect(content).toMatchSnapshot()
})

test('if internal is set to false module should not be marked as internal ', async () => {
    await handler({
        source: "fixtures/notInternalModule",
        outFile: 'fixtures/notInternalModule/CssModules.fs',
        internal: false,
        _: [],
        $0: ''
    })

    const content = readFileSync("fixtures/notInternalModule/CssModules.fs").toString()
    expect(content).toMatchSnapshot()
})

test('module name with a dot should be transformed to a valid F# identifier', async () => {
    await handler({
        source: "fixtures/moduleWithDot",
        outFile: 'fixtures/moduleWithDot/CssModules.fs',
        internal: true,
        _: [],
        $0: ''
    })

    const content = readFileSync("fixtures/moduleWithDot/CssModules.fs").toString()
    expect(content).toMatchSnapshot()
})

test('nested folder is supported', async () => {
    await handler({
        source: "fixtures/nestedFolder",
        outFile: 'fixtures/nestedFolder/CssModules.fs',
        internal: true,
        _: [],
        $0: ''
    })

    const content = readFileSync("fixtures/nestedFolder/CssModules.fs").toString()
    expect(content).toMatchSnapshot()
})

test("nested CSS is supported", async () => {
    await handler({
        source: "fixtures/nestedCss",
        outFile: 'fixtures/nestedCss/CssModules.fs',
        internal: true,
        _: [],
        $0: ''
    })

    const content = readFileSync("fixtures/nestedCss/CssModules.fs").toString()
    expect(content).toMatchSnapshot()
})
