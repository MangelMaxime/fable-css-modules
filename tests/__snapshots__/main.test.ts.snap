// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`contains 1 class if there is one CSS Module 1`] = `
"//------------------------------------------------------------------------------
//        This code was generated by fable-css-modules.
//        Changes to this file will be lost when the code is regenerated.
//------------------------------------------------------------------------------

[<RequireQualifiedAccess>]
module internal CssModules

open Fable.Core

type App =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

"
`;

exports[`contains 1 class per CSS Module 1`] = `
"//------------------------------------------------------------------------------
//        This code was generated by fable-css-modules.
//        Changes to this file will be lost when the code is regenerated.
//------------------------------------------------------------------------------

[<RequireQualifiedAccess>]
module internal CssModules

open Fable.Core

type App =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

type Main =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

"
`;

exports[`if internal is set to false module should not be marked as internal  1`] = `
"//------------------------------------------------------------------------------
//        This code was generated by fable-css-modules.
//        Changes to this file will be lost when the code is regenerated.
//------------------------------------------------------------------------------

[<RequireQualifiedAccess>]
module CssModules

open Fable.Core

type App =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

"
`;

exports[`module name with a dot should be transformed to a valid F# identifier 1`] = `
"//------------------------------------------------------------------------------
//        This code was generated by fable-css-modules.
//        Changes to this file will be lost when the code is regenerated.
//------------------------------------------------------------------------------

[<RequireQualifiedAccess>]
module internal CssModules

open Fable.Core

type User_Index =

    /// <summary>
    /// Binding for <c>list</c> class
    /// </summary>
    [<Emit("$0[\\"list\\"]")>]
    abstract list : string

"
`;

exports[`nested CSS is supported 1`] = `
"//------------------------------------------------------------------------------
//        This code was generated by fable-css-modules.
//        Changes to this file will be lost when the code is regenerated.
//------------------------------------------------------------------------------

[<RequireQualifiedAccess>]
module internal CssModules

open Fable.Core

type CSS =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

    /// <summary>
    /// Binding for <c>blue</c> class
    /// </summary>
    [<Emit("$0[\\"blue\\"]")>]
    abstract blue : string

type SASS =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

    /// <summary>
    /// Binding for <c>blue</c> class
    /// </summary>
    [<Emit("$0[\\"blue\\"]")>]
    abstract blue : string

type SCSS =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

    /// <summary>
    /// Binding for <c>blue</c> class
    /// </summary>
    [<Emit("$0[\\"blue\\"]")>]
    abstract blue : string

"
`;

exports[`nested folder is supported 1`] = `
"//------------------------------------------------------------------------------
//        This code was generated by fable-css-modules.
//        Changes to this file will be lost when the code is regenerated.
//------------------------------------------------------------------------------

[<RequireQualifiedAccess>]
module internal CssModules

open Fable.Core

type App =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

module Folder1 =

    type App =

        /// <summary>
        /// Binding for <c>red</c> class
        /// </summary>
        [<Emit("$0[\\"red\\"]")>]
        abstract red : string

    module Folder1_1 =

        type App =

            /// <summary>
            /// Binding for <c>red</c> class
            /// </summary>
            [<Emit("$0[\\"red\\"]")>]
            abstract red : string

"
`;

exports[`sanitize class name 1`] = `
"//------------------------------------------------------------------------------
//        This code was generated by fable-css-modules.
//        Changes to this file will be lost when the code is regenerated.
//------------------------------------------------------------------------------

[<RequireQualifiedAccess>]
module internal CssModules

open Fable.Core

type App =

    /// <summary>
    /// Binding for <c>user-name</c> class
    /// </summary>
    [<Emit("$0[\\"user-name\\"]")>]
    abstract userName : string

    /// <summary>
    /// Binding for <c>user__age</c> class
    /// </summary>
    [<Emit("$0[\\"user__age\\"]")>]
    abstract user__age : string

"
`;

exports[`support CSS/SASS/SCSS modules 1`] = `
"//------------------------------------------------------------------------------
//        This code was generated by fable-css-modules.
//        Changes to this file will be lost when the code is regenerated.
//------------------------------------------------------------------------------

[<RequireQualifiedAccess>]
module internal CssModules

open Fable.Core

type CSS =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

type SASS =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

type SCSS =

    /// <summary>
    /// Binding for <c>red</c> class
    /// </summary>
    [<Emit("$0[\\"red\\"]")>]
    abstract red : string

"
`;
